x-pack/plugins/viz_editor/server/query/select_operations.ts

```diff
@@ -0,0 +1,232 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+/*********************************************************************************************
+ * Logic to convert our query shape into an Elastic search query. It takes a flat list
+ * of selection criteria, and converts them into the various-- sometimes heirardchical--
+ * shapes expected by the Elasticsearch DSL.
+ *********************************************************************************************/
+
+import {
+  AvgOperation,
+  chunkBy,
+  ChunkResult,
+  Column,
+  CountOperation,
+  DateHistogramOperation,
+  isEmpty,
+  partition,
+  Query,
+  SelectOperation,
+} from '../../common';
+
+/**
+ * A handful of transformation functions for a select operation,
+ * used to build up an ES query and flatten the results.
+ */
+export interface SelectDefinition {
+  toEsAgg?: (op: any) => any;
+  toNestedQuery?: (query: Query, ops: SelectOperation[], getSubAggs: () => any) => any;
+  getName?: (op: any) => string;
+}
+
+/**
+ * A map of select operations (col, date_histogram, count, etc) to their
+ * SelectDefinition, used to convert our definitions into the equivalent
+ * Elasticsearch DSL.
+ */
+export const selectOperations: { [operation: string]: SelectDefinition } = {
+  col: {
+    getName(op: Column) {
+      return op.alias || op.argument;
+    },
+    toNestedQuery(query: Query, ops: SelectOperation[], getSubAggs: () => any) {
+      const cols = ops as Column[];
+      return {
+        aggregations: {
+          groupby: {
+            composite: {
+              size: query.size,
+              sources: cols.map(col => ({
+                [col.alias!]: {
+                  terms: {
+                    field: col.argument,
+                    missing_bucket: true,
+                    order: 'asc',
+                  },
+                },
+              })),
+            },
+            ...getSubAggs(),
+          },
+        },
+      };
+    },
+  },
+
+  date_histogram: {
+    getName(op: DateHistogramOperation) {
+      return op.alias || `${op.argument.field}_${op.argument.interval}`;
+    },
+    toNestedQuery(query: Query, ops: SelectOperation[], getSubAggs: () => any) {
+      if (ops.length > 1) {
+        throw new Error(`Multiple date_histograms are not supported.`);
+      }
+
+      const op = ops[0] as DateHistogramOperation;
+
+      return {
+        aggregations: {
+          [op.alias!]: {
+            date_histogram: op.argument,
+            ...getSubAggs(),
+          },
+        },
+      };
+    },
+  },
+
+  count: {
+    getName(op: AvgOperation) {
+      return op.alias || 'count';
+    },
+    toEsAgg(op: CountOperation) {
+      // We perform a count in this way so that it
+      // is consistently represented regardless of
+      // other query concerns (groupby, histogram, etc)
+      return {
+        [op.alias!]: {
+          value_count: {
+            field: '_id',
```

Comment:
I'm not sure I follow your reasoning for avoiding the count aggregation, can you explain?

Reply:
There isn't a count aggregation (that I'm aware of). Instead, count is one of several properties, depending on the query: `hits.total.value` or `aggregations.groupby.buckets[i].doc_count`, etc. Computing count with `value_count` makes this always consistent, regardless of the query. Not sure if this is a good idea, but it made parsing the results simpler.